File: perf/benchmarks/micro/xor_shift/xor_shift.py
# xor_shift — xorshift64 PRNG tight-loop micro-benchmark.
#
# Iterate xorshift64 N times starting from a fixed seed and return the final
# state. Pure 64-bit bitwise / shift ops. Argument: N (default 50_000_000).

import sys
import time

MASK = 0xFFFFFFFFFFFFFFFF


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 50_000_000
    t0 = time.perf_counter_ns()
    s = 0x123456789ABCDEF0
    for _ in range(n):
        s ^= (s << 13) & MASK
        s ^= (s >> 7) & MASK
        s ^= (s << 17) & MASK
        s &= MASK
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{s}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/micro/xor_shift/xor_shift.cpp
// xor_shift — xorshift64 PRNG tight-loop micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 50000000;
    auto t0 = std::chrono::steady_clock::now();
    uint64_t s = 0x123456789ABCDEF0ULL;
    for (int64_t i = 0; i < n; ++i) {
        s ^= s << 13;
        s ^= s >> 7;
        s ^= s << 17;
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%llu\n", static_cast<unsigned long long>(s));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
